dowhy/gcm/util/catboost_encoder.py

```diff
@@ -0,0 +1,130 @@
+from typing import Optional
+
+import numpy as np
+from sklearn.preprocessing import LabelEncoder
+
+
+class CatBoostEncoder:
+    """Implements the proposed method from
+
+    "CatBoost: gradient boosting with categorical features support", Dorogush et al. (2018).
+
+    The Catboost encoder is a target encoder for categorical features. In this implementation we follow Eq. (1)
+    in https://arxiv.org/pdf/1810.11363.pdf.
+    """
+
+    def __init__(self, p: float = 1, alpha: Optional[float] = None):
+        """See Eq. (1) in https://arxiv.org/pdf/1810.11363.pdf
+
+        :param p: The p parameter in the equation. This weights the impact of the given alpha.
+        :param alpha: Alpha parameter in the equation. If None is given, the global mean is used as suggested in
+                      "A preprocessing scheme for high-cardinality categorical attributes in classification and
+                      prediction problems", Micci-Barreca (2001)
+        """
+        self._p = p
+        self._org_alpha = alpha
+        self._category_means = None
+
+    def _fit_transform(self, X: np.ndarray, Y: np.ndarray, use_alpha_when_unique: bool, train: bool) -> np.ndarray:
```

Comment:
nit: please move this method to the bottom.

Reply:
done